### DISCUSSION TITLE
Failure vs Error in regards to RetryPolicy

### DISCUSSION THREAD

**Discussion by jmccrumm** [argoproj/argo-workflows#10429] [Q&A]
Where is the line drawn between an Error and a Failure when it comes to the [retryPolicy](https://argoproj.github.io/argo-workflows/walk-through/retrying-failed-or-errored-steps/) for a Retry Strategy?

I ask because we seem to have a lot of intermittent problems with some of our workflows connecting to our internal GitLab instance, in which case a hit of the 'Retry' button in Argo Workflows always solves it. I attempted to put some retry strategies in our workflows so that we don't have to manually retry when these intermittent issues occur, but I set the `retryPolicy` to `OnError`, and it seems there are some cases that are considered `Failure` when I would think they'd be `Error`s. For instance, we get `artifact <blah> failed to load...ssh: handshake failed: EOF`, but this is marked as a `Failure`, when it didn't even get to executing the actual script for that step. I could set the `retryPolicy` to `Always`, but I also don't want it to keep retrying when there is a true failure when running the script itself. 

Also, what is `OnTransientError`? If that case occurs, will the `Phase` within the workflow show exactly that?

Thanks.

I am running v3.4.4 of Argo Workflows.

**Comment by wesleyscholl** [comment]
@jmccrumm 

**Failure** refers to scenarios where the main container of a step executes but terminates with a non-zero exit code, indicating that the process encountered an issue during execution. The default retryPolicy, **OnFailure**, retries steps whose main container is marked as failed in Kubernetes. 

**Error**, refers to issues that prevent the step from executing its main task. Problems like failures in init or wait containers, or workflow controller errors. The **OnError** retryPolicy retries steps that encounter such errors. 

In your situation, where intermittent Gitlab issues result in errors like: “artifact <blah> failed to load...ssh: handshake failed: EOF”, should be classified as Errors because the main container hasn’t executed the primary script. Therefore, setting the retryPolicy to **OnError** should, in theory, address these issues.

However, if you’re observing errors are marked as Failures, it might be due to how the system interprets specific error conditions.

The **OnTransientError**, policy retries steps that encounter errors defined as transient. Examples include I/O or TLS handshake timeouts. These are classified as a subset of errors that are primarily temporary and may succeed by retrying. If a step fails with a transient error, the workflow’s phase will reflect this, and the step will be retried if the **OnTransientError** retryStrategy has been defined.


**Resources:**

https://argo-workflows.readthedocs.io/en/release-3.4/retries

https://github.com/argoproj/argo-workflows/blob/main/examples/retry-on-error.yaml

**Reply by kenchan0130** [reply]
Does this mean that the pod will not be recreated unless the policy is set to `OnError`?
When the policy was set to `OnFailure`, the workflow stopped without retrying, so I would like to confirm this.

https://argo-workflows.readthedocs.io/en/latest/tolerating-pod-deletion/

**Reply by valayDave** [reply]
Thank you so much for taking the time to answer this question. I am a core dev on the [Netflix/Metaflow](https://github.com/Netflix/metaflow) project, which uses Argo Workflows to deploy machine learning workflows on Kubernetes.

I’m currently working through a somewhat complicated use case that I haven't found a good workaround for yet. When deploying workflows, we allow users to set retry behavior through Python decorators. We then translate that configuration down to the Argo workflow template. 

At the moment, we do not distinguish between *errors* and *failures* — retries apply universally because we set the `retryPolicy` to `Always`. However, what we really want is a way to differentiate between the two. This is difficult because, in the current `expr` context, we only have access to `lastRetry`, which doesn’t provide a complete picture (e.g., how many past retries were due to errors versus failures). 

If there were a way to access some form of global state — specifically, a count of retries due to errors versus failures — we could craft an expression that ensures user-configured retries apply purely to failures. This would allow us, as library designers, to provide sensible defaults and configuration knobs for handling retries on errors separately.
